File: old_version/Python/第4章 贪心算法/4.2 活动安排/activity_selector.py
def select_activity(a, s, f):
    activity = []
    n = len(s)
    activity.append(a[0])
    k = 0
    for i in range(1, n):
        if s[i] >= f[k]:
            activity.append(a[i])
            k = i
    return activity

def main():
    # 活动集合
    a = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]
    # 活动开始时间集合
    s = [1, 3, 0, 5, 3, 5, 6, 8, 8, 2, 12]
    # 活动结束时间集合
    f = [4, 5, 6, 7, 9, 9, 10, 11, 12, 14, 16]
    print("选择活动：%s" % select_activity(a, s, f))

if __name__ == "__main__":
    main()

File: code/C/Chapter11/11-2/activity_selector.c
#include <stdio.h>
#include <stdlib.h>

int* selectActivity(int *a, int aSize, 
               int *s, int sSize, 
               int *f, int fSize, int *activityCnt) {
    int *activity = (int *)malloc(sizeof(int) * aSize);
    if(!activity) {
        return NULL;
    }

    *activityCnt = 0;
    activity[(*activityCnt)++] = a[0];
    int k = 0;
    for(int i = 1; i < sSize; i++) {
        if(s[i] >= f[k]) {
            activity[(*activityCnt)++] = a[i];
            k = i;
        }
    }

    return activity;
}

int main() {
    // 活动集合
    int a[] = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11};
    int aSize = sizeof(a) / sizeof(a[0]);
    // 活动开始时间集合
    int s[] = {1, 3, 0, 5, 3, 5, 6, 8, 8, 2, 12};
    int sSize = sizeof(s) / sizeof(s[0]);
    // 活动结束时间集合
    int f[] = {4, 5, 6, 7, 9, 9, 10, 11, 12, 14, 16};
    int fSize = sizeof(f) / sizeof(f[0]);

    int cnt = 0;        // 选择的活动数量
    int *activity = selectActivity(a, aSize, s, sSize, f, fSize, &cnt);
    if(activity) {
        printf("选择活动：");
        for(int i = 0; i < cnt; i++) {
            printf("%d ", activity[i]);
        }
        printf("\n");
        free(activity);
    }

    return 0;
}